README.md
# Cb Response Event Forwarder

## Overview

The Cb Response Event Forwarder is a standalone service that will listen on the Cb Response enterprise bus and export
events (both watchlist/feed hits as well as raw endpoint events, if configured) in a normalized JSON or LEEF format.
The events can be saved to a file, delivered to a network service or archived automatically to an Amazon AWS S3 bucket.
These events can be consumed by any external system that accepts JSON or LEEF, including Splunk and IBM QRadar.

##Configuration 

The 4.0 Cb Response Event Forwarder has a three part processing pipeline which is specified:
in YAML format - with two crucicial parts - input and output.

1) input - this section defines a number of consumers each reading from a specific CbR server messaging bus
The list of events to collect is configurable.
By default all feed and watchlist hits, alerts, binary notifications, and raw sensor events are exported into JSON.  The
```
input: 
    cbserver1:
        cb_server_url: https://zestep-centos-cbresponseserver 
        # rabbitmq options are opitional, /etc/cb/.conf is used by default
        # if the folling are ommited
        rabbit_mq_username: cb
        rabbit_mq_hostname: localhost
        rabbit_mq_password:  
```
#Use this boolean option to control the use of the raw sensor exchange
#defaults to false
`use_raw_exchange : true`

## Subscribed Events 
The 4.0 event forwarder subcribes to `all-events` by default though the user
can specify explicit options for which events to listen to as so:
(within each individual input element)
event_map:
    events_watchlist:
        - watchlist.#
    events_feed:
        - feed.#
    events_alert:
        - alert.#
    events_raw_sensor:
        - ingress.event.process 
        - ingress.event.procstart
        - ingress.event.netconn
        - ingerss.event.procend
        - ingress.event.childproc
        - ingress.event.moduleload
        - ingress.event.module
        - ingress.event.filemod
        - ingress.event.regmod
        - ingress.event.tamper
        - ingress.event.crossprocopen
        - ingress.event.remotethread
        - ingress.event.processblock
        - ingress.event.emetmitigation
    events_binary_observed:
        - binaryinfo.#
    events_binary_upload:    
        - binarystore.#
    events_storage_parition:
        - events.partition.#

AMQPs tls can be configured by providing a `tls:` stanza within an individual input: 
```
tls:    
    client_cert: yourcert
    client_key: yourkey
    ca_cert: cacert
    tls_insecure: true # defaults false 
```

There are optional post processing arguments to enhance messages w/ api
callbacks (currently just report information)
```
post_processing:
    tls: 
        #slightly different than other TLS stanzas found in input/output
        verify: false # defaults to true
        tls_12_only: false # defaults to true 
        client_cert: client.cert
        client_key: client.key
        ca_cert: ca.cert
    api_proxy_url: proxyurl
    api_token: apitoken
```
Just specify the post processing options within an individual input element.
    
2) filter - this section of the pipeline defines an optional template for keeping/droping messages based on their contents. 
This is a top-level option, along side input: and output: 
```
filter:
    template: >-
              {{if (eq .type "alert.watchlist.hit.query.binary") -}}
                KEEP
              {{- else -}}
                DROP
              {{- end}}
```

3) output - this section defines a number of outputs, which can be of differing types (file,socket, etc)  and formats (json,LEEF, custom)
```
output:
    - file:
        path: "output-json.txt"
        format:
            type: json 
    - file:
        path: "output-leef.txt"
        format:
            type: leef 
    - file:
        path: "output-yaml.txt"
        format:
            type: template 
            template: "{{YamlFormat .}}"
```
## Supported Output Types in 4.0
The 4.0.0 CbR event-forwarder supports the same output options as 3.X:
Each output must provide path/connection information, other configuration options
specific to that output type and a format (like the old `output_format`) like so:

output formats
```
format: 
    type: json # leef, cef, template are the possible options
```
output options
```
file: 
    path: path/to/your/desiredoutput.type
    format:
        type: json
http:
    destination: https://myserver:51337
    # overide the default template ... http_post_template: 
    #upload_empty_files: false    
    #bundle_size_max: "50000000"
    #bundle_send_timeout: "500"
    #comma_seperate_events: false 
    # defaults to false 
    #headers:
    #   header: headervalue
    tls:
        #tls options
    format:
        type: json
socket:
   connection: tpc://host-or-ip:port 
   format: 
        type: json
splunk:
syslog:
    connection: tcp://host-or-ip:port 

```

## Support

The pre-built RPM is supported via our [User eXchange (Jive)](https://community.carbonblack.com/community/developer-relations) 
and via email to dev-support@carbonblack.com.  

## Raw Sensor Events 

We have seen a performance impact when exporting all raw sensor events onto the enterprise bus.  We do not recommend
exporting all the events.  The performance impacts are seen when the events are broadcast on the bus, by enabling the
"DatastoreBroadcastEventTypes".  We recommend that at most, only process and netconn events be broadcast on the event
bus. 

## Quickstart Guide

The cb-event-forwarder can be installed on any 64-bit Linux machine running CentOS 6.x. 
It can be installed on the same machine as the Cb Response server, or another machine. 
If you are forwarding a large volume of events to QRadar (for example, all file modifications and/or registry 
modifications), or are forwarding events from a Cb Response cluster, then installing it on a separate machine is recommended. 
Otherwise, it is acceptable to install the cb-event-forwarder on the Cb Response server itself.

### Installation

To install and configure the cb-event-forwarder, perform these steps as "root" on your target Linux system.

1. Install the CbOpenSource repository if it isn't already present:

    ```
    cd /etc/yum.repos.d
    curl -O https://opensource.carbonblack.com/release/x86_64/CbOpenSource.repo
    ```
2. Install the RPM via YUM: 

   ```
   yum install cb-event-forwarder
   ```

### Configure the cb-event-forwarder

Note: The 4.X event forwarder uses Yaml files for configuration. Values will be interpreted as their literal types, rather than using 0,1,T/F to represent booleans `true` or `false` should be used.

1. If installing on a machine *other than* the Cb Response server, copy the RabbitMQ username and password into the 
`rabbit_mq_username` and `rabbit_mq_password` variables in `/etc/cb/integrations/event-forwarder/cb-event-forwarder.conf` 
file. Also fill out the `cb_server_hostname` with the hostname or IP address where the Cb Response server can be reached.
If the cb-event-forwarder is forwarding events from a Cb Response cluster, the `cb_server_hostname` should be set
to the hostname or IP address of the Cb Response master node.
   
2. Ensure that the configuration is valid by running the cb-event-forwarder in Check mode: 
`/usr/share/cb/integrations/event-forwarder/cb-event-forwarder -check` as root. If everything is OK, you will see a 
message starting with "Initialized output”. If there are any errors, those errors will be printed to your screen.

### Configure Cb Response

By default, Cb publishes the `feed.*` and `watchlist.*` events over the bus (see the [Events documentation](EVENTS.md)
for more information). 
If you want to capture raw sensor events or the `binaryinfo.*` notifications, you have to enable those features in 
`/etc/cb/cb.conf`:

* If you are capturing raw sensor events then you also need to edit the `DatastoreBroadcastEventTypes` option in 
`/etc/cb/cb.conf` to enable broadcast of the raw sensor events you wish to export.
* If you are capturing binary observed events you also need to edit the `EnableSolrBinaryInfoNotifications` option in 
`/etc/cb/cb.conf` and set it to true.

Cb Response needs to be restarted if any variables were changed in `/etc/cb/cb.conf` by executing
`service cb-enterprise restart`. 

If you are configuring the cb-event-forwarder on a Cb Response cluster, the `DatastoreBroadcastEventTypes` and/or
`EnableSolrBinaryInfoNotifications` settings
must be distributed to the `/etc/cb/cb.conf` configuration file on all minion nodes and the cluster stopped and started using
the `/usr/share/cb/cbcluster stop && /usr/share/cb/cbcluster start` command.

### Starting and Stopping the Service

Once the service is installed, it is managed by the Upstart init system in CentOS 6.x. You can control the service via the 
initctl command.
* To start the service, `initctl start cb-event-forwarder`
* To stop the service, `initctl stop cb-event-forwarder`

Once the service is installed, it is configured to start automatically on system boot.

## Splunk

The Cb Response event forwarder can be used to export Cb Response events in a way easily configured for Splunk.  You'll
need to install and configure the Splunk TA to consume the Cb Response event data.   It is recommended that the event
bridge use a file based output with Splunk universal forwarder configured to monitor the file.   

More information about configuring the Splunk TA can be found [here](http://docs.splunk.com/Documentation/AddOns/latest/Bit9CarbonBlack/About)

## QRadar

The Cb Response event forwarder can forward Cb Response events in the LEEF format to QRadar. To forward Cb Response
events to a QRadar server:

1. Modify `/etc/cb/integrations/event-forwarder/cb-event-forwarder.conf` to include 
`udpout: "<qradaripaddress>:<port>"` (NOTE: Port is usually 514)
2. Change the output format to LEEF in the configuration file: `output_format: leef`.
3. Change the output type to UDP in the configuration file: `output_type: udp`.

For more information on the LEEF format, see the [Events documentation](EVENTS.md).

## Logging & Diagnostics

The connector logs to the directory `/var/log/cb/integrations/cb-event-forwarder`. An example of a successful startup log:

```
2015/12/07 12:57:26 cb-event-forwarder version 4.0.0 starting
2015/12/07 12:57:26 Interface address 172.22.10.7
2015/12/07 12:57:26 Interface address fe80::20c:29ff:fe85:bcd0
2015/12/07 12:57:26 Configured to capture events: [watchlist.hit.# watchlist.storage.hit.# feed.ingress.hit.# 
feed.storage.hit.# feed.query.hit.# alert.watchlist.hit.# ingress.event.process ingress.event.procstart 

CHANGES.md
# Changelog

## cb-event-forwader 4.0.0 beta

The 4.0.0 release of the cb-event-forwarder retains the features of 3x and: 
- enhanced architecture - specify multiple CbR servers for input and multiple outputs 
- enhanced configuration - configuration format changed from .ini to yaml 
- templated outputs format - users can marshal output messages in arbitrary formats using golang templates
    - Fully compatible with existing outputs**
- output plugins - users can write their own output plugins to support output types not supported currently, w/o contributing code upstream (but we hope you still do) 
    - The plugin architecture is meant for advanced users and is subject to artibtrary change as the CbEF evolves
- filtering - Users can filter events during processing in the event forwarder, specifing a keep-or-drop template in the relevant config section
    - filtering happens to all events, from all inputs
- enhancements to existing outputs
    - more graceful management of temporary files by BundledOutputs ( splunk-hec, s3, http)
- Kafa Output changed to output-plugin
    - 'standard-plugin' will be provided with the CbR EF
        - makes a wonderful template for other output plugins 
    - the kafka-output has been enhanced :
        - uses librdkafka/go-confluent-kafka instead of sarama
            -better protocol compliance/support
        - users can now specify arbitrary kafka-producer configs (at their own peril - consider how your broker(s) are configured too) 


## cb-event-forwarer 3.5.0
- kafka SASL support
- OATH2 JWT optional support for http output
- Support for sending EventText as bytearary httpoutput 

## cb-event-forwarder 3.1.2

The 3.1.2 release of cb-event-forwarder adds two features:
 
* You can now send arbitrary messages for debugging/testing purposes through the forwarder to the output location.
  This is only available when the cb-event-forwarder is started with the `-debug` command line switch. Messages
  sent via this mechanism are also logged for audit purposes.
* S3: You can now explicitly specify the location of the AWS credential file to use for authentication in the
  `credential_profile` option in the `[s3]` section of the configuration file. To search for the credential profile
  `production` in the credentials stored in the file `/etc/cb/aws.creds`, set the `credential_profile` option to
  `/etc/cb/aws.creds:production`.

## cb-event-forwarder 3.1.1

The 3.1.1 release of cb-event-forwarder fixes a critical bug when rolling over files. Previous versions of the
cb-event-forwarder would stop rolling over files after the first of a new month. This release fixes that bug.

## cb-event-forwarder 3.1.0

The 3.1.0 release of cb-event-forwarder adds the following features over 3.0.0:

* "Deep links" into the Cb server UI are now optionally available in the output
  * These links allow you to directly access the relevant sensor, binary, or process context for each event output
    by the cb-event-forwarder.
  * The new variable `cb_server_url` has been added to the configuration file to support this new feature. Set this
    variable to the base URL of the Carbon Black web UI. **If this variable is not set, then no links are generated.**
  * The new links are available in the `link_process`, `link_child` (in child process events), `link_md5` and 
    `link_sensor` keys of the JSON or LEEF output.
  * Note that links to processes and binaries may result in 404 errors until the process and binary data is committed
    to disk on the Carbon Black server. Process events received via the event-forwarder may take up to 15 minutes or 
    longer before they're visible on the Carbon Black web UI.
* All Carbon Black 5.1 event types are now supported
  * Microsoft EMET
  * Carbon Black Tamper events
  * Cross-process (process open/thread create) events
  * Carbon Black process/network blocking events
* Network events now include the local IP and port number of the network connection (available on Carbon Black 5.1 
  servers and sensors)
  * The IP four-tuple is now available as (`local_ip`, `local_port`, `remote_ip`, and `remote_port`) in the JSON/LEEF
    output
* Provide a human-readable status page for statistics
  * By default, these statistics are available via HTTP on port 33706 of the system running the cb-event-forwarder.
* Fix regressions on output from cb-event-forwarder 2.x on some JSON message types
  * cb-event-forwarder 3.0.0 was missing the `computer_name` field from some JSON messages
* New Amazon S3 options; see the `[s3]` section of the configuration file
  * Specify whether the files uploaded to S3 should be encrypted with server-side encryption (see `server_side_encryption`)
  * Define an ACL policy to apply to files uploaded to S3 (see `acl_policy`)
  * Specify the credential profile used when connecting to S3 (see `credential_profile`)

# Changes from the cb-event-forwarder 2.x to 3.x

In general, the new cb-event-forwarder 3.0 is designed to be a drop-in replacement for previous versions of the
event forwarder. There are a few bug fixes, configuration changes and enhancements of note. The most important change
is that the service is now managed by the "upstart" system in CentOS 6. The `service` command is no longer used to

config.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	_ "expvar"
	"fmt"
	log "github.com/sirupsen/logrus"
	"github.com/vaughan0/go-ini"
	"io/ioutil"
	"strconv"
	"strings"
	"text/template"
	"time"
)

const (
	FileOutputType = iota
	S3OutputType
	TCPOutputType
	UDPOutputType
	SyslogOutputType
	HTTPOutputType
	SplunkOutputType
	KafkaOutputType
)

const (
	LEEFOutputFormat = iota
	JSONOutputFormat
)

type Configuration struct {
	ServerName           string
	AMQPHostname         string
	DebugFlag            bool
	DebugStore           string
	OutputType           int
	OutputFormat         int
	AMQPUsername         string
	AMQPPassword         string
	AMQPPort             int
	AMQPTLSEnabled       bool
	AMQPTLSClientKey     string
	AMQPTLSClientCert    string
	AMQPTLSCACert        string
	AMQPQueueName        string
	AMQPAutomaticAcking  bool
	OutputParameters     string
	EventTypes           []string
	EventMap             map[string]bool
	HTTPServerPort       int
	CbServerURL          string
	UseRawSensorExchange bool

	// this is a hack for S3 specific configuration
	S3ServerSideEncryption  *string
	S3CredentialProfileName *string
	S3ACLPolicy             *string
	S3ObjectPrefix          *string

	// SSL/TLS-specific configuration
	TLSClientKey  *string
	TLSClientCert *string
	TLSCACert     *string
	TLSVerify     bool
	TLSCName      *string
	TLS12Only     bool

	// HTTP-specific configuration
	HTTPAuthorizationToken *string
	HTTPPostTemplate       *template.Template
	HTTPContentType        *string

	// configuration options common to bundled outputs (S3, HTTP)
	UploadEmptyFiles    bool
	CommaSeparateEvents bool
	BundleSendTimeout   time.Duration
	BundleSizeMax       int64

	// Compress data on S3 or file output types
	FileHandlerCompressData bool

	TLSConfig *tls.Config

	// optional post processing of feed hits to retrieve titles
	PerformFeedPostprocessing bool
	CbAPIToken                string
	CbAPIVerifySSL            bool
	CbAPIProxyURL             string

	// Kafka-specific configuration
	KafkaBrokers     *string
	KafkaTopicSuffix *string

	//Splunkd
	SplunkToken *string

	RemoveFromOutput []string
	AuditLog         bool
}

type ConfigurationError struct {
	Errors []string
	Empty  bool
}

func (c *Configuration) AMQPURL() string {
	return fmt.Sprintf("amqp://%s:%s@%s:%d", c.AMQPUsername, c.AMQPPassword, c.AMQPHostname, c.AMQPPort)
}

func (e ConfigurationError) Error() string {
	return fmt.Sprintf("Configuration errors:\n %s", strings.Join(e.Errors, "\n "))
}

func (e *ConfigurationError) addErrorString(err string) {
	e.Empty = false
	e.Errors = append(e.Errors, err)
}

func (e *ConfigurationError) addError(err error) {
	e.Empty = false
	e.Errors = append(e.Errors, err.Error())
}

func parseCbConf() (username, password string, err error) {
	input, err := ini.LoadFile("/etc/cb/cb.conf")
	if err != nil {
		return username, password, err
	}
	username, _ = input.Get("", "RabbitMQUser")
	password, _ = input.Get("", "RabbitMQPassword")

	if len(username) == 0 || len(password) == 0 {
		return username, password, errors.New("Could not get RabbitMQ credentials from /etc/cb/cb.conf")
	}
	return
}

func (c *Configuration) parseEventTypes(input ini.File) {
	eventTypes := [...]struct {
		configKey string
		eventList []string
	}{
		{"events_watchlist", []string{
			"watchlist.#",
		}},
		{"events_feed", []string{
			"feed.#",
		}},
		{"events_alert", []string{
			"alert.#",
		}},
		{"events_raw_sensor", []string{
			"ingress.event.process",
			"ingress.event.procstart",
			"ingress.event.netconn",
			"ingress.event.procend",
			"ingress.event.childproc",
			"ingress.event.moduleload",
			"ingress.event.module",
			"ingress.event.filemod",
			"ingress.event.regmod",
			"ingress.event.tamper",
			"ingress.event.crossprocopen",
			"ingress.event.remotethread",
			"ingress.event.processblock",
			"ingress.event.emetmitigation",
		}},
		{"events_binary_observed", []string{
			"binaryinfo.#",
		}},
		{"events_binary_upload", []string{
			"binarystore.#",
		}},
		{"events_storage_partition", []string{
			"events.partition.#",
		}},
	}

	for _, eventType := range eventTypes {
		val, ok := input.Get("bridge", eventType.configKey)
		if ok {
			val = strings.ToLower(val)
			if val == "all" {
				for _, routingKey := range eventType.eventList {
					c.EventTypes = append(c.EventTypes, routingKey)
				}
			} else if val == "0" {
				// nothing
			} else {
				for _, routingKey := range strings.Split(val, ",") {
					c.EventTypes = append(c.EventTypes, routingKey)
				}
			}
		}
	}

	c.EventMap = make(map[string]bool)

	log.Info("Raw Event Filtering Configuration:")
	for _, eventName := range c.EventTypes {
		c.EventMap[eventName] = true
		if strings.HasPrefix(eventName, "ingress.event.") {
			log.Infof("%s: %t", eventName, c.EventMap[eventName])
		}
	}
}

func ParseConfig(fn string) (Configuration, error) {
	config := Configuration{}
	errs := ConfigurationError{Empty: true}

	input, err := ini.LoadFile(fn)
	if err != nil {
		return config, err
	}

	// defaults
	config.DebugFlag = false
	config.OutputFormat = JSONOutputFormat
	config.OutputType = FileOutputType
	config.AMQPHostname = "localhost"
	config.AMQPUsername = "cb"
	config.HTTPServerPort = 33706
	config.AMQPPort = 5004
	config.DebugStore = "/tmp"

	config.S3ACLPolicy = nil
	config.S3ServerSideEncryption = nil
	config.S3CredentialProfileName = nil

	// required values
	val, ok := input.Get("bridge", "server_name")
	if !ok {
		config.ServerName = "CB"
	} else {
		config.ServerName = val
	}

	val, ok = input.Get("bridge", "debug")
	if ok {
		if val == "1" {
			config.DebugFlag = true
			log.SetLevel(log.DebugLevel)

			customFormatter := new(log.TextFormatter)
			customFormatter.TimestampFormat = "2006-01-02 15:04:05"
			log.SetFormatter(customFormatter)
			customFormatter.FullTimestamp = true

			log.Debug("Debugging output is set to True")
		}
	}

	removeFromOutput, ok := input.Get("bridge", "remove_from_output")
	if ok {
		thingsToRemove := strings.Split(removeFromOutput, ",")
		numberOfThingsToRemove := len(thingsToRemove)
		strippedThingsToRemove := make([]string, numberOfThingsToRemove)
		for index, element := range thingsToRemove {
			strippedThingsToRemove[index] = strings.TrimSpace(element)
		}
		if numberOfThingsToRemove > 0 {
			config.RemoveFromOutput = strippedThingsToRemove
		} else {
			config.RemoveFromOutput = make([]string, 0)
		}
	} else {
		config.RemoveFromOutput = make([]string, 0)
	}

	debugStore, ok := input.Get("bridge", "debug_store")
	if ok {
		config.DebugStore = debugStore
	} else {
		config.DebugStore = "/var/log/cb/integrations/cb-event-forwarder"
	}

	log.Debugf("Debug Store is %s", config.DebugStore)

	val, ok = input.Get("bridge", "http_server_port")
	if ok {
		port, err := strconv.Atoi(val)
		if err == nil {
			config.HTTPServerPort = port
		}
	}

	val, ok = input.Get("bridge", "rabbit_mq_username")
	if ok {
		config.AMQPUsername = val
	}

	val, ok = input.Get("bridge", "rabbit_mq_password")
	if !ok {
		errs.addErrorString("Missing required rabbit_mq_password section")
	} else {
		config.AMQPPassword = val
	}

	val, ok = input.Get("bridge", "rabbit_mq_port")
	if ok {
		port, err := strconv.Atoi(val)
		if err == nil {
			config.AMQPPort = port
		}
	}

	if len(config.AMQPUsername) == 0 || len(config.AMQPPassword) == 0 {
		config.AMQPUsername, config.AMQPPassword, err = parseCbConf()
		if err != nil {
			errs.addError(err)
		}
	}

	val, ok = input.Get("bridge", "rabbit_mq_use_tls")
	if ok {
		if ok {
			b, err := strconv.ParseBool(val)
			if err == nil {
				config.AMQPTLSEnabled = b
			}
		}
	}

	rabbitKeyFilename, ok := input.Get("bridge", "rabbit_mq_key")
	if ok {
		config.AMQPTLSClientKey = rabbitKeyFilename
	}

	rabbitCertFilename, ok := input.Get("bridge", "rabbit_mq_cert")
	if ok {
		config.AMQPTLSClientCert = rabbitCertFilename
	}

	rabbitCaCertFilename, ok := input.Get("bridge", "rabbit_mq_ca_cert")
	if ok {
		config.AMQPTLSCACert = rabbitCaCertFilename
	}

	rabbitQueueName, ok := input.Get("bridge", "rabbit_mq_queue_name")
	if ok {
		config.AMQPQueueName = rabbitQueueName
	}

	config.AMQPAutomaticAcking = true
	rabbitAutomaticAcking, ok := input.Get("bridge", "rabbit_mq_automatic_acking")
	if ok {
		boolval, err := strconv.ParseBool(rabbitAutomaticAcking)
		if err == nil {
			if boolval == false {
				config.AMQPAutomaticAcking = false
			}
		} else {
			errs.addErrorString("Unknown value for 'rabbit_mq_automatic_acking': valid values are true, false, 1, 0. Default is 'true'")
		}
	}

	val, ok = input.Get("bridge", "cb_server_hostname")
	if ok {
		config.AMQPHostname = val
	}

	val, ok = input.Get("bridge", "cb_server_url")
	if ok {
		if !strings.HasSuffix(val, "/") {
			val = val + "/"
		}
		config.CbServerURL = val
	}

	val, ok = input.Get("bridge", "output_format")
	if ok {
		val = strings.TrimSpace(val)
		val = strings.ToLower(val)
		if val == "leef" {
			config.OutputFormat = LEEFOutputFormat
		}
	}

	config.FileHandlerCompressData = false
	val, ok = input.Get("bridge", "compress_data")
	if ok {
		b, err := strconv.ParseBool(val)
		if err == nil {
			config.FileHandlerCompressData = b
		}
	}

	config.AuditLog = false
	val, ok = input.Get("bridge", "audit_log")
	if ok {
		b, err := strconv.ParseBool(val)
		if err == nil {
			config.AuditLog = b
		}
	}

	outType, ok := input.Get("bridge", "output_type")
	var parameterKey string
	if ok {
		outType = strings.TrimSpace(outType)
		outType = strings.ToLower(outType)

		switch outType {
		case "file":
			parameterKey = "outfile"
			config.OutputType = FileOutputType
		case "tcp":
			parameterKey = "tcpout"
			config.OutputType = TCPOutputType
		case "udp":
			parameterKey = "udpout"
			config.OutputType = UDPOutputType
		case "s3":
			parameterKey = "s3out"
			config.OutputType = S3OutputType

			profileName, ok := input.Get("s3", "credential_profile")
			if ok {
				config.S3CredentialProfileName = &profileName
			}

			aclPolicy, ok := input.Get("s3", "acl_policy")
			if ok {
				config.S3ACLPolicy = &aclPolicy
			}

			sseType, ok := input.Get("s3", "server_side_encryption")
			if ok {
				config.S3ServerSideEncryption = &sseType
			}

			objectPrefix, ok := input.Get("s3", "object_prefix")
			if ok {
				config.S3ObjectPrefix = &objectPrefix
			}

		case "http":
			parameterKey = "httpout"
			config.OutputType = HTTPOutputType

			token, ok := input.Get("http", "authorization_token")
			if ok {
				config.HTTPAuthorizationToken = &token
			}

			postTemplate, ok := input.Get("http", "http_post_template")
			config.HTTPPostTemplate = template.New("http_post_output")
			if ok {
				config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(postTemplate))
			} else {
				if config.OutputFormat == JSONOutputFormat {
					config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(
						`{"filename": "{{.FileName}}", "service": "carbonblack", "alerts":[{{range .Events}}{{.EventText}}{{end}}]}`))
				} else {
					config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(`{{range .Events}}{{.EventText}}{{end}}`))
				}
			}

			contentType, ok := input.Get("http", "content_type")
			if ok {
				config.HTTPContentType = &contentType
			} else {
				jsonString := "application/json"
				config.HTTPContentType = &jsonString
			}
		case "syslog":
			parameterKey = "syslogout"
			config.OutputType = SyslogOutputType
		case "kafka":
			config.OutputType = KafkaOutputType

			kafkaBrokers, ok := input.Get("kafka", "brokers")
			if ok {
				config.KafkaBrokers = &kafkaBrokers
			}

			kafkaTopicSuffix, ok := input.Get("kafka", "topic_suffix")
			if ok {
				config.KafkaTopicSuffix = &kafkaTopicSuffix
			}
		case "splunk":
			parameterKey = "splunkout"
			config.OutputType = SplunkOutputType

			token, ok := input.Get("splunk", "hec_token")
			if ok {
				config.SplunkToken = &token
			}

			postTemplate, ok := input.Get("splunk", "http_post_template")
			config.HTTPPostTemplate = template.New("http_post_output")
			if ok {
				config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(postTemplate))
			} else {
				if config.OutputFormat == JSONOutputFormat {
					config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(
						`{{range .Events}}{"sourcetype":"bit9:carbonblack:json","event":{{.EventText}}}{{end}}`))
				} else {
					config.HTTPPostTemplate = template.Must(config.HTTPPostTemplate.Parse(`{{range .Events}}{{.EventText}}{{end}}`))
				}
			}

			contentType, ok := input.Get("http", "content_type")
			if ok {
				config.HTTPContentType = &contentType
			} else {
				jsonString := "application/json"
				config.HTTPContentType = &jsonString
			}

		default:
			errs.addErrorString(fmt.Sprintf("Unknown output type: %s", outType))
		}
	} else {
		errs.addErrorString("No output type specified")
		return config, errs
	}

	if len(parameterKey) > 0 {
		val, ok = input.Get("bridge", parameterKey)
		if !ok {
			errs.addErrorString(fmt.Sprintf("Missing value for key %s, required by output type %s",
				parameterKey, outType))
		} else {
			config.OutputParameters = val
		}
	}

	val, ok = input.Get("bridge", "use_raw_sensor_exchange")
	if ok {
		boolval, err := strconv.ParseBool(val)
		if err == nil {
			config.UseRawSensorExchange = boolval
			if boolval {
				log.Warn("Configured to listen on the Carbon Black Enterprise Response raw sensor event feed.")
				log.Warn("- This will result in a *large* number of messages output via the event forwarder!")
				log.Warn("- Ensure that raw sensor events are enabled in your Cb server (master & minion) via")
				log.Warn("  the 'EnableRawSensorDataBroadcast' variable in /etc/cb/cb.conf")
			}
		} else {
			errs.addErrorString("Unknown value for 'use_raw_sensor_exchange': valid values are true, false, 1, 0")
		}
	}

	// TLS configuration
	clientKeyFilename, ok := input.Get(outType, "client_key")
	if ok {
		config.TLSClientKey = &clientKeyFilename
	}

	clientCertFilename, ok := input.Get(outType, "client_cert")
	if ok {
		config.TLSClientCert = &clientCertFilename
	}

	caCertFilename, ok := input.Get(outType, "ca_cert")
	if ok {
		config.TLSCACert = &caCertFilename
	}

	config.TLSVerify = true
	tlsVerify, ok := input.Get(outType, "tls_verify")
	if ok {
		boolval, err := strconv.ParseBool(tlsVerify)
		if err == nil {
			if boolval == false {
				config.TLSVerify = false
			}
		} else {
			errs.addErrorString("Unknown value for 'tls_verify': valid values are true, false, 1, 0. Default is 'true'")
		}
	}

	config.TLS12Only = true
	tlsInsecure, ok := input.Get(outType, "insecure_tls")
	if ok {
		boolval, err := strconv.ParseBool(tlsInsecure)
		if err == nil {
			if boolval == true {
				config.TLS12Only = false
			}
		} else {
			errs.addErrorString("Unknown value for 'insecure_tls': ")
		}
	}

	serverCName, ok := input.Get(outType, "server_cname")
	if ok {
		config.TLSCName = &serverCName
	}

	config.TLSConfig = configureTLS(config)

	// Bundle configuration

	// default to sending empty files to S3/HTTP POST endpoint
	if outType == "splunk" {
		config.UploadEmptyFiles = false
		log.Info("Splunk HEC does not accept empty files as input, ignoring upload_empty_files=true for 'splunkout'")
	} else {
		config.UploadEmptyFiles = true
	}
	sendEmptyFiles, ok := input.Get(outType, "upload_empty_files")
	if ok {
		boolval, err := strconv.ParseBool(sendEmptyFiles)
		if err == nil {
			if boolval == false {
				config.UploadEmptyFiles = false
			}
		} else {
			errs.addErrorString("Unknown value for 'upload_empty_files': valid values are true, false, 1, 0. Default is 'true'")
		}
	}

	if config.OutputFormat == JSONOutputFormat {
		config.CommaSeparateEvents = true
	} else {
		config.CommaSeparateEvents = false
	}

	// default 10MB bundle size max before forcing a send
	config.BundleSizeMax = 10 * 1024 * 1024
	bundleSizeMax, ok := input.Get(outType, "bundle_size_max")
	if ok {
		bundleSizeMax, err := strconv.ParseInt(bundleSizeMax, 10, 64)
		if err == nil {
			config.BundleSizeMax = bundleSizeMax
		}
	}

	// default 5 minute send interval
	config.BundleSendTimeout = 5 * time.Minute
	bundleSendTimeout, ok := input.Get(outType, "bundle_send_timeout")
	if ok {
		bundleSendTimeout, err := strconv.ParseInt(bundleSendTimeout, 10, 64)
		if err == nil {
			config.BundleSendTimeout = time.Duration(bundleSendTimeout) * time.Second
		}
	}

	val, ok = input.Get("bridge", "api_verify_ssl")
	if ok {
		config.CbAPIVerifySSL, err = strconv.ParseBool(val)
		if err != nil {
			errs.addErrorString("Unknown value for 'api_verify_ssl': valid values are true, false, 1, 0. Default is 'false'")
		}
	}
	val, ok = input.Get("bridge", "api_token")
	if ok {
		config.CbAPIToken = val
		config.PerformFeedPostprocessing = true
	}

	config.CbAPIProxyURL = ""
	val, ok = input.Get("bridge", "api_proxy_url")
	if ok {
		config.CbAPIProxyURL = val
	}

	config.parseEventTypes(input)

	if !errs.Empty {
		return config, errs
	}
	return config, nil
}

func configureTLS(config Configuration) *tls.Config {
	tlsConfig := &tls.Config{}

	if config.TLSVerify == false {
		log.Info("Disabling TLS verification for remote output")
		tlsConfig.InsecureSkipVerify = true
	}

	if config.TLSClientCert != nil && config.TLSClientKey != nil && len(*config.TLSClientCert) > 0 &&
		len(*config.TLSClientKey) > 0 {
		log.Infof("Loading client cert/key from %s & %s", *config.TLSClientCert, *config.TLSClientKey)
		cert, err := tls.LoadX509KeyPair(*config.TLSClientCert, *config.TLSClientKey)
		if err != nil {
			log.Fatal(err)
		}
		tlsConfig.Certificates = []tls.Certificate{cert}
	}

	if config.TLSCACert != nil && len(*config.TLSCACert) > 0 {
		// Load CA cert
		log.Infof("Loading valid CAs from file %s", *config.TLSCACert)
		caCert, err := ioutil.ReadFile(*config.TLSCACert)
		if err != nil {
			log.Fatal(err)
		}
		caCertPool := x509.NewCertPool()
		caCertPool.AppendCertsFromPEM(caCert)
		tlsConfig.RootCAs = caCertPool
	}

	if config.TLSCName != nil && len(*config.TLSCName) > 0 {
		log.Infof("Forcing TLS Common Name check to use '%s' as the hostname", *config.TLSCName)
		tlsConfig.ServerName = *config.TLSCName
	}

	if config.TLS12Only == true {
		log.Info("Enforcing minimum TLS version 1.2")
		tlsConfig.MinVersion = tls.VersionTLS12
	} else {
		log.Info("Relaxing minimum TLS version to 1.0")
		tlsConfig.MinVersion = tls.VersionTLS10
	}

	return tlsConfig
}
